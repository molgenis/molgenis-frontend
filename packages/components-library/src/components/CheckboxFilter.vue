<template>
  <div>
    <b-form-checkbox-group
      v-model="selection"
      stacked
      :options="visibleOptions"
    />
    <span v-if="bulkOperation">
      <b-link
        v-if="showToggleSlice"
        class="toggle-slice card-link"
        @click.prevent="toggleSlice"
      >
        {{ toggleSliceText }}
      </b-link>
      <b-link
        class="toggle-select card-link"
        @click.prevent="toggleSelect"
      >
        {{ toggleSelectText }}
      </b-link>
    </span>
  </div>
</template>

<style>
  .card-link { font-style: italic; font-size: small; }
</style>

<script>
export default {
  props: {
    name: {
      type: String,
      required: true
    },
    label: {
      type: String,
      required: false,
      default: () => ''
    },
    options: {
      type: [Function],
      required: true
    },
    value: {
      type: Array,
      default: () => []
    },
    bulkOperation: {
      type: Boolean,
      required: false,
      default: () => true
    },
    maxVisibleOptions: {
      type: Number,
      default: () => undefined
    }
  },
  data () {
    return {
      resolvedOptions: [],
      sliceOptions: this.maxVisibleOptions && this.resolvedOptions && this.maxVisibleOptions < this.resolvedOptions.length
    }
  },
  computed: {
    selection: {
      get () {
        return this.value
      },
      set (value) {
        this.$emit('input', value.length === 0 ? undefined : value)
      }
    },
    visibleOptions () {
      return this.sliceOptions ? this.resolvedOptions.slice(0, this.maxVisibleOptions) : (typeof this.resolvedOptions === 'function' ? [] : this.resolvedOptions)
    },
    showToggleSlice () {
      return this.maxVisibleOptions && this.maxVisibleOptions < this.resolvedOptions.length
    },
    toggleSelectText () {
      return this.value.length ? 'Deselect all' : 'Select all'
    },
    toggleSliceText () {
      return this.sliceOptions ? `Show ${this.resolvedOptions.length - this.maxVisibleOptions} more` : 'Show less'
    }
  },
  watch: {
    resolvedOptions () {
      this.sliceOptions = this.showToggleSlice
    },
    maxVisibleOptions () {
      this.sliceOptions = this.showToggleSlice
    }
  },
  created () {
    this.options().then(response => {
      this.resolvedOptions = response
    })
  },
  methods: {
    toggleSelect () {
      this.selection = this.selection && this.selection.length ? [] : this.resolvedOptions.map(option => option.value)
    },
    toggleSlice () {
      this.sliceOptions = !this.sliceOptions
    }
  }
}
</script>

<docs>
Filter that renders a list of options as a set of checkboxes

## Examples

```jsx
Basic checkbox:

const options = () => Promise.resolve(
 [
  { text: 'Orange', value: 'orange' },
  { text: 'Apple', value: 'apple' },
  { text: 'Pineapple', value: 'pineapple' },
  { text: 'Grape', value: 'grape' }
 ]
)
const selection = []
<checkbox-filter
  name="example-checkbox"
  v-bind:options="options"
  v-model="selection">
</checkbox-filter>

<div>Selected</div>
<ul><li v-for="item in selection">{{item}}</li></ul>

```
</docs>
