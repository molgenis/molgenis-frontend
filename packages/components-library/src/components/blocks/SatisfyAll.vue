<template>
  <div class="text-right mb-2 satisfy-all-checkbox">
    <label>
      {{ satisfyAllLabel }}
      <b-form-checkbox
        v-model="satisfyAllModel"
        name="satisfy-all"
        class="d-inline-block ml-1"
      />
    </label>
  </div>
</template>

<script>
export default {
  props: {
    /**
     * This is the satisfyAll v-model value. 
     * @model
     */
    
    value: {
      type: Boolean,
      default: () => false
    },
    /**
     * The label to show on the left of the satisfy all Checkbox
     */
    satisfyAllLabel: {
      type: String,
      required: false,
      default: () => 'Satisfy all'
    }
  },
  computed: {
    satisfyAllModel: {
      get () {
        return this.value
      },
      set (value) {
        this.$emit('input', value)
      }
    }
  }
}
</script>

<docs>
Filter that renders a list of options as a set of checkboxes

## Usage
```jsx
const satisfyAllLabel = 'Satisfy all'
let satisfyAll = false
    <satisfy-all
      v-bind:satisfy-all-label="satisfyAllLabel"
      v-model="satisfyAll"
    />
<div>satisfyAll: {{satisfyAll}}</div>
```
</docs>
