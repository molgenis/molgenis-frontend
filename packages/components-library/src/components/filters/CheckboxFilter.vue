<template>
  <div>
    <b-form-checkbox-group
      v-model="selection"
      stacked
      :options="visibleOptions"
    />
    <span v-if="bulkOperation">
      <b-link
        v-if="showToggleSlice"
        class="toggle-slice card-link"
        @click.prevent="toggleSlice"
      >
        {{ toggleSliceText }}
      </b-link>
      <b-link
        class="toggle-select card-link"
        @click.prevent="toggleSelect"
      >
        {{ toggleSelectText }}
      </b-link>
    </span>
  </div>
</template>

<script>
export default {
  name: 'CheckboxFilter',
  props: {
    /**
     * Toggle to switch between returning an array with values or an array with the full option
     */
    returnTypeAsObject: {
      type: Boolean,
      required: false,
      default: () => false
    },
    /**
     * A Promise-function that resolves with an array of options.
     * {text: 'foo', value: 'bar'}
     */
    options: {
      type: [Function],
      required: true
    },
    /**
     * This is the v-model value; an array of selected options.
     * Can also be a { text, value } object array
     */
    value: {
      type: Array,
      default: () => []
    },
    /**
     * Whether to use (De)Select All or not.
     */
    bulkOperation: {
      type: Boolean,
      required: false,
      default: () => true
    },
    /**
     * Limit the maximum number of visible items.
     */
    maxVisibleOptions: {
      type: Number,
      default: () => undefined
    }
  },
  data () {
    return {
      externalUpdate: false,
      selection: [],
      resolvedOptions: [],
      sliceOptions: this.maxVisibleOptions && this.resolvedOptions && this.maxVisibleOptions < this.resolvedOptions.length
    }
  },
  computed: {
    visibleOptions () {
      return this.sliceOptions ? this.resolvedOptions.slice(0, this.maxVisibleOptions) : (typeof this.resolvedOptions === 'function' ? [] : this.resolvedOptions)
    },
    showToggleSlice () {
      return this.maxVisibleOptions && this.maxVisibleOptions < this.resolvedOptions.length
    },
    toggleSelectText () {
      return this.value.length ? 'Deselect all' : 'Select all'
    },
    toggleSliceText () {
      return this.sliceOptions ? `Show ${this.resolvedOptions.length - this.maxVisibleOptions} more` : 'Show less'
    }
  },
  watch: {
    value () {
      this.setValue()
    },
    resolvedOptions () {
      this.sliceOptions = this.showToggleSlice
    },
    selection (newValue) {
      if (!this.externalUpdate) {
        let newSelection = []

        if (this.returnTypeAsObject) {
          newSelection = Object.assign(newSelection, this.resolvedOptions.filter(of => newValue.includes(of.value)))
        } else {
          newSelection = [...newValue]
        }
        this.$emit('input', newSelection)
      }
      this.externalUpdate = false
    }
  },
  created () {
    this.options().then(response => {
      this.resolvedOptions = response
    })
    this.setValue()
  },
  methods: {
    toggleSelect () {
      if (this.selection && this.selection.length > 0) {
        this.selection = []
      } else {
        this.selection = this.resolvedOptions.map(option => option.value)
      }
    },
    toggleSlice () {
      this.sliceOptions = !this.sliceOptions
    },
    setValue () {
      this.externalUpdate = true
      if (this.value && this.value.length > 0 && typeof this.value[0] === 'object') {
        this.selection = this.value.map(vo => vo.value)
      } else {
        this.selection = this.value
      }
    }
  }
}
</script>

<style>
.card-link { font-style: italic; font-size: small; }
</style>

<docs>
Filter that renders a list of options as a set of checkboxes

## Usage
```jsx
const model = []
<CheckboxFilter
  v-bind:maxVisibleOptions="5"
  v-bind:bulkOperation="true"
  v-bind:options="fruitOptionsFunction"
  v-model="model">
</CheckboxFilter>
<div>model: {{model}}</div>
```
</docs>
