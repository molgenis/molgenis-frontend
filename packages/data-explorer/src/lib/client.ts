import build from './clientFactory'
import store from '@/store/store'
import { AxiosInstance } from 'axios'

export const errorReponse = (error: any) => {
  let message = error.message
  if (error.response && error.response.data && error.response.data.detail) {
    message = error.response.data.detail
  }
  if (error.response && error.response.status === 401) {
    if (!store.getters.isUserAuthenticated) {
      window.location.href = '/login'
    }
  }
  // Set timeout to '0' to keep visible until user dismisses the error notification
  store.commit('addToast', { message, type: 'danger', timeout: 0 })
  return Promise.reject(error)
}

// Create the default axios client the data-explorer will use
const client:AxiosInstance = build({
  responseErrorInterceptor: errorReponse
})

export default client
