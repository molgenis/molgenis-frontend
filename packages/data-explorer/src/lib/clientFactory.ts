/* eslint-disable no-console */
import axios, { AxiosInstance, AxiosResponse } from 'axios'
import { ClientSettings } from '@/types/ClientSettings'

export const defaultResponse = function (response: AxiosResponse<any>) {
  return response
}

export const defaultErrorReponse = function (error: any) {
  if (console && console.error) {
    console.error(`${error.errorCode}: ${error.title}`)
    console.error(`${error.detail}`)
  }
  return Promise.reject(error)
}

export const defaultStatusValidation = function (status: number) {
  return status >= 200 && status < 300
}

export function build (settings?: ClientSettings):AxiosInstance {
  const client:AxiosInstance = axios.create({
    baseURL: settings && settings.baseURL ? settings.baseURL : '/',
    timeout: settings && settings.timeout ? settings.timeout : 5000,
    validateStatus: settings && settings.validateStatus ? settings.validateStatus : defaultStatusValidation
  })

  client.interceptors.response.use(
    settings && settings.responseInterceptor ? settings.responseInterceptor : defaultResponse,
    settings && settings.responseErrorInterceptor ? settings.responseErrorInterceptor : defaultErrorReponse
  )

  return client
}

export default build
