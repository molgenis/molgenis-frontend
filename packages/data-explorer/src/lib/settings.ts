import Axios from 'axios'
import store from '../store/store'

function getSettingsUrl () {
  return `${window.location.href.replace(window.location.hash, '').replace('index.html', '')}config/settings.xlsx`
}

function setSettingsFormData (fileContents: string) {
  let formData = new FormData()
  let dataAsBlob = new Blob([fileContents], { type: 'application/xml' })
  formData.append('file', dataAsBlob, 'settings.xlsx')
  formData.append('action', 'ADD_UPDATE_EXISTING')
  formData.append('notify', 'false')
  return formData
}

function settingsByUrlFormdata () {
  return `/plugin/importwizard/importByUrl?notify=true&entityTypeId=demo&url=${encodeURIComponent(getSettingsUrl())}`
}

const BootstrapExplorer = async () => {
  let versionNumberRequest = await Axios.get('/config/version.txt')
  const version = versionNumberRequest.data
  // check for table
  // if not then check user rights
  // const fileRequest = await Axios.get('/config/settings.xlsx')
  // const settingsFormData = setSettingsFormData(fileRequest.data)
  // console.log(fileRequest)
  const t = settingsByUrlFormdata()
  let x = await Axios.post(t)

  // /plugin/importwizard/importByUrl?notify=false&entityTypeId=demo&url=http://localhost:8080/config/settings.xlsx
  if (store.getters.hasEditRights) { }

  // Axios.get('/config/settingstable.xlsx').then(t => console.log(t))
  let isSettingsTableAvailable = Axios.get('/api/metadata/root_hospital_patients?flattenAttributes=true').then(t =>
    console.log(t)
  )
}

export default BootstrapExplorer
