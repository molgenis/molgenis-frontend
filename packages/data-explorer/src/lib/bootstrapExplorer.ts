import client from './client'
import store from '../store/store'
import applicationSettings from './applicationSettings'

const packageEndpoint = 'api/data/sys_md_Package'
const metaDataEndpoint = 'api/metadata'

async function createSettings () {
  if (!store.getters.hasEditRights) {
    return {
      type: 'danger',
      message: 'Please login as administrator to initialize the application'
    }
  }

  try {
    await client.post(packageEndpoint, applicationSettings.packageSettings)
  } catch (error) {
    // if the settings table already exists, axios gives error and stops executing.
  }

  await client.post(metaDataEndpoint, applicationSettings.entitySettings)
  return {
    type: 'success',
    message: 'The application has been succesfully initialized!'
  }
}

const bootstrapExplorer = async (): Promise<any | undefined> => {
  try {
    await client.get('/api/data/app_set_DataExplorerEntitySettings')
  } catch (error) {
    store.commit('setToast', { type: 'warning', message: 'Bootstrapping Data explorer' })
    const status = await createSettings()
    store.commit('setToast', { type: status.type, message: status.message })
  }
}

export default bootstrapExplorer
