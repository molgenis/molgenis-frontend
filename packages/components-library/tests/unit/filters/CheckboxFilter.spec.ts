import { mount } from '@vue/test-utils'
import CheckboxFilter from '@/components/filters/CheckboxFilter.vue'
import SatisfyAll from '@/components/blocks/SatisfyAll.vue'
import { localVue as getLocalVue } from '../../lib/helpers'

const localVue = getLocalVue()

function getWrapper(props = {}) {
  const propsData = {
    ...props,
    value: [],
    options: async () => {
      return [
        { value: 'foo', text: 'Foo' },
        { value: 'bar', text: 'Bar' },
        { value: 'baz', text: 'Baz' }
      ]
    }
  }

  const wrapper = mount(CheckboxFilter, {
    localVue,
    components: { SatisfyAll },
    stubs: ['font-awesome-icon'],
    propsData,
    listeners: {
      input: (newVal: any) => {
        wrapper.setProps({ value: newVal })
      }
    }
  })
  return wrapper
}

describe('CheckboxFilter.vue', () => {
  describe('Parsing only required props', () => {
    let wrapper: any
    beforeEach(() => { wrapper = getWrapper({}) })

    it('can set and unset values', async () => {
      const inputElements = wrapper.findAll('input')
      inputElements.at(0).trigger('click') // select foo
      inputElements.at(1).trigger('click') // select bar
      inputElements.at(2).trigger('click') // select baz
      inputElements.at(1).trigger('click') // deselect bar
      await localVue.nextTick()
      expect(wrapper.emitted('input')[0]).toEqual([['foo', 'baz']])
    })

    it('can return objects', async () => {
      wrapper.setProps({ returnTypeAsObject: true })

      const inputElements = wrapper.findAll('input')
      inputElements.at(0).trigger('click') // select foo
      inputElements.at(1).trigger('click') // select bar
      await localVue.nextTick()
      expect(wrapper.emitted('input')[0]).toEqual([[{ value: 'foo', text: 'Foo' },
      { value: 'bar', text: 'Bar' }]])
    })

    it('can select all and deselect all', async () => {
      const toggleSelect = wrapper.find('.toggle-select')
      expect(toggleSelect.text()).toBe('Select all')
      await toggleSelect.trigger('click') // select all
      expect(toggleSelect.text()).toBe('Deselect all')
      expect(wrapper.findAll('input[type=checkbox]:checked').length).toEqual(3)
      await toggleSelect.trigger('click') // deselect all
      expect(wrapper.findAll('input[type=checkbox]:checked').length).toEqual(0)
    })
  })

  describe('When restricting maxVisibleOptions to 1', () => {
    let wrapper: any
    beforeEach(() => { wrapper = getWrapper({ maxVisibleOptions: 1 }) })

    it('can hide elements based on maxVisibleOptions', async () => {
      expect(wrapper.findAll('input[type=checkbox]').length).toBe(1)

      const sliceButton = wrapper.find('.toggle-slice')
      await sliceButton.trigger('click')
      expect(wrapper.findAll('input[type=checkbox]').length).toBe(3)
      await sliceButton.trigger('click')
      expect(wrapper.findAll('input[type=checkbox]').length).toBe(1)
    })
  })

  describe('SatisfyAllCheckbox', () => {
    it('checks that the satisfyAll checkbox is not shown by default ', () => {
      const wrapper = getWrapper()
      const satisfyAllButton = wrapper.find('input[name="satisfy-all"]')
      expect(satisfyAllButton.exists()).toBe(false)
    })

    it('triggers the proper emit when the satisfyAll checkbox is clicked', async () => {
      const wrapper = getWrapper({ showSatisfyAllCheckbox: true })
      const satisfyAllButton = wrapper.find('input[name="satisfy-all"]')
      await satisfyAllButton.trigger('click')
      expect(wrapper.emitted('satisfy-all')).toEqual([[true]])
    })
  })
})
