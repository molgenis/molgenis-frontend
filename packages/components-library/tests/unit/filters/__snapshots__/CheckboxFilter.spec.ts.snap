// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxFilter.vue Parsing only required props matches the snapshot 1`] = `
<div>
  <div
    class="bv-no-focus-ring"
    id="__BVID__4"
    role="group"
    tabindex="-1"
  >
    <div
      class="custom-control custom-checkbox"
    >
      <input
        class="custom-control-input"
        id="__BVID__4_BV_option_0"
        name="__BVID__4"
        type="checkbox"
        value="foo"
      />
      <label
        class="custom-control-label"
        for="__BVID__4_BV_option_0"
      >
        <span>
          Foo
        </span>
      </label>
    </div>
    <div
      class="custom-control custom-checkbox"
    >
      <input
        class="custom-control-input"
        id="__BVID__4_BV_option_1"
        name="__BVID__4"
        type="checkbox"
        value="bar"
      />
      <label
        class="custom-control-label"
        for="__BVID__4_BV_option_1"
      >
        <span>
          Bar
        </span>
      </label>
    </div>
    <div
      class="custom-control custom-checkbox"
    >
      <input
        class="custom-control-input"
        id="__BVID__4_BV_option_2"
        name="__BVID__4"
        type="checkbox"
        value="baz"
      />
      <label
        class="custom-control-label"
        for="__BVID__4_BV_option_2"
      >
        <span>
          Baz
        </span>
      </label>
    </div>
  </div>
   
  <span>
    <!---->
     
    <a
      class="toggle-select card-link"
      href="#"
      target="_self"
    >
      
      Select all
    
    </a>
  </span>
</div>
`;
